### DISCUSSION TITLE
SDF driver for Windows version of Visit

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#406] [General]
Hi, 
Visit or Paraview can load hell amount (~200) of different file formats but not the EPOCH's SDF one. Hence EPOCH provides additional driver for specific version of Visit. Unfortunately there is no version of Visit for my version of Linux, compilation of Visit from GetHub took hours and failed, Visit folks are helpless with this version of Linux too for the last two years, and for Windows version of Visit it is probably harder to create the driver because all MAKE files there are different. But may be anyone has succeeded to make SDF driver for Visit in Windows?

Alternatively can anyone help to get the particles and fields saved to harddrive directly from EPOCH avoiding these FOUR (!!!) intermediate "black box" stages: 

saving to SDF --> SDF DRIVER converts SDF file for Visit --> Visit loads data --> Visit dumps the data in ASCII/binary format. 

Not even mentioning that we do not need these Visit, Paraview or any other graphics software nice pics, we just need the raw data. 
I have made the beta version for this saving but it messes output somewhere. It works only if no MPI parallelization is used or essentially only if one core is used. When say two cores are used the final data does not represent the sum of two parts to make the whole entire simulation box, the pieces of data get distorted somewhere like in a wrong assembled puzzle. Does each core work not with real x,y,z coordinates but with the additionally shifted ones ?

It is completely understandable that the decision to use SDF, HDF etc file formats were taken years ago when there was no compatible way to save and load binary data and saving just the text ASCII is slow and inefficient. But with latest additions to Standard  ACCESS = "STREAM" all binary formats now became compatible along all existing compilers and OSes. Plus this way to read/write is 10-100x faster, infinitely simpler and files are ~2x smaller than with SDF

**Comment by Status-Mirror** [comment]
Hi Dan,

Working directly with the SDF binary is complicated, although I believe some FORTRAN and C++ code is present in the SDF directory.

I also struggle to get Visit working on Linux, and I find the least-effort approach is to use the MATLAB or Python functions for extracting and manipulating SDF data. These are also present in the SDF directory (Python is labelled as utils).

My personal preference is to use the MATLAB's GetDataSDF("0000.sdf") function to get the data contained in the 0000.sdf file. Alternatively, Python users may wish to follow the instructions documented here:
https://epochpic.github.io/documentation/visualising_output/python.html

We are about to start a large rewrite of EPOCH to scale with next generation super-computers, and the I/O will likely be updated here. We appreciate the suggestion of using STREAM - this is worth looking into. I wouldn't expect any big changes any time soon though, the new EPOCH project is a 3 year grant.

Cheers,
Stuart

**Reply by DanRRRR** [reply]
Thanks Stuart for respond. Great that you got the grant for so needed development and improvements. That gives the hope that the activity here will be rising. 
I think that chosen by the developers decision to use Fortran for EPOCH and not any other language was great right design decision. I sometimes do not believe how insanely fast EPOCH is due to that. 

That will overweight all other not so great minor design choices like taking SDF format for input/output while there was not a single one with its own Fortran I/O, and while Fortran READ/WRITE facility is way more superior to many other languages in speed, simplicity and reliability. Sometimes even plain ASCII would be enough for getting some data quick. 100x100x100 or 1000x1000 arrays are not that large nowadays. And the speed of NVMe drives is crazily fast today. I wrote above that to get the data from SDF takes 4 awkward black box stages how counterintuitive this format is. But i forgot one more - users will needs also to install Python or MATLAB to get at the end 100-1000x slower Python output from EPOCH :)))). Also why MATLAB or Python? Some Fortran compilers are sometimes better even to make all the plots than MATLAB or Python, look at Silverfrost Fortran for example with their native graphics library, native PL with just one line and native OpenGL all done in GUI.

That negative design decision would  not be considered as bad decision but as one of more features of EPOCH if there was a Fortran way to output data. Can developers make this quick fix for the  users to stop suffering with SDF? The developers acknowledge that by themselves : "The low level Python SDF library is not user friendly, so a wrapper layer called sdf_helper has been written....". 

The simple ASCII/binary (set in initial file)  Fortran output of  x,y,z, px,py,pz, weight, z, and ID is the only what  needed for each sort of particles and in the same format the x,y,z,Ex,Ey,Ez for the fields (for H and J too).  

Also, please correct me if i am wrong, but  really large design defect of EPOCH is the total absence of lost particles outputs even in SDF format. That means that sometimes EPOCH is not usable for comparisons with some real experiments. It is like you would characterize solar radiation emission by just modeling  what is going on inside the sun while the experiment of course observes the radiation from outside the simulation box in our case called the "Sun" and the photons that escape  this simulation box are exactly what the experiment observe. So, creating output for lost particles is also absolutely needed.  

It's ok, and even more user friendly and network transfer friendly to have separate Fortran data files for all these outputs instead of one huge pile of everything with SDF.

**Reply by Status-Mirror** [reply]
We use SDF for it's parallel I/O capabiliites, which is a format developed here at Warwick. The main benefit of this approach is its scalability - it is tried, tested and able to write output dumps of hundreds of GB from thousands of processors. This is a common requirement for large-scale 2D simulations. I think it's unfair to say users are "suffering with SDF" - the MATLAB implementation takes one line to load and sort all the data from a single output dump (GetDataSDF). Python is a free alternative, which only takes a few more steps to get running.

We will likely switch to an even more accessible file-type in the EPOCH re-write (perhaps HDF), but we have no plans to develop the existing EPOCH outputs. The I/O system in EPOCH would be difficult to change in its current state. However, this is a project with an open source code, so someone could add an ASCII output to the code if they wanted to. Currently data is written to file in lines like "write_particle_variable" in diagnostics.F90. The determined coder could adapt this function to write to ASCII, instead of SDF. Although, I believe it would take less effort to learn SDF.

If "lost particle outputs" refers to particles leaving the simulation window, you can output these. You'll want to add "ejected_particles" to your output block, as discussed here: https://epochpic.github.io/documentation/input_deck/input_deck_output_block.html

**Reply by TomGoffrey** [reply]
I'd also add there's existing tools, shipped with EPOCH that will read SDF and output in plain text. You could do something like `./sdffilter ../../epoch1d/Data/0000.sdf -c > test.txt` from the SDF/utilities directory for example.

**Reply by DanRRRR** [reply]
Stuart,
Now we have 100s of formats, some of them were not accepted as broadly as expected like with SDF, some eventually getting abandoned due to their own reasons . The same with HDF, people started to abandon it, see "Moving away from HDF5" https://cyrille.rossant.net/moving-away-hdf5/. Trying to support all these formats Visit reached the limit that to compile it takes whole day and it can not succeed with the latest versions for Ubintu for the last two years :). Similar app like Visit 100% written in Fortran +OpenGL takes 5 seconds to compile with Silverfrost Fortran compiler

Currently parallel capabilities have little to add with Fortran as the inherent bandwidth of Fortran I/O is saturating the bandwidth of any latest hardware and even  in serial case beats some parallel formats. Bandwidth of NVMe gen 4.0 for example is 8 GB per second and Fortran gives crazy 3-7 GB /second even with Windows PC (6GB/s Intel, 3GB/s gFortran, 7GB/s  Silverfrost) . Those who do not believe me, try to create 4-10GB real*4 array say Arr(11,200000000) and check how much time it will take on NVMe 4.0 to write. The complete code for that is here

real*4 Arr(11,200000000)
Arr(:,:)=1.2345
Open(unit=30,file='test' ,access='stream', form='unformatted')
WRITE(30) Arr
close(30)
end

It takes 1-2 seconds! That is how easy is to save 100% portable way all EPOCH variables. Just 3 lines OPEN-WRITE-CLOSE. And the same to load. Your SDF format took >1 Magabyte of just Fortran sources in EPOCH code and additional drivers for Visit (!!!) Crazy.

I have nothing against SDF if there existed the standard Fortran output of major, basic 10 particle (coordinates, momentum, weight and ID) and 9 field variables for B,E,J. Synthetic variables like temperature, density and many others provided with "write_particle_variable" including possibility to output even some graphics plots are extra i am not even interested with.   

This input/output is roughest part for anyone starting with EPOCH. Even with MATLAB it will take time to set it up. It took me several months to find the way how to avoid learning one more almost dead cumbersome black box format. And i am not even done yet.

**Reply by TomGoffrey** [reply]
A few things to consider:

1. How does your proposed use of raw Fortran output scale to 100k+ processes? Assuming you stick to the current output strategy that'll result in billions of files for a typical simulation? How do you propose to handle that in post-processing? This sort of output strategy may well be problematic for your filesystem too.
2. You seem to imply that 3-7 GB/s is going to out-perform parallel I/O. I don't think this is true. Taking ARCHER (last generation UK HPC) as an example, they achieved something around 16,000MiB/s during benchmarking (around 50% of theoretical peak performance of the file system). This was 6 years ago, and ARCHER has since been decommissioned.
3. If you want to ignore the first 2 points above, as Stuart says EPOCH is open source under a permissive license. It wouldn't be too hard to write your own output routines.
4. As I mentioned above, tools are shipped with EPOCH to translate SDF into plain text, should you want to use that.

**Reply by DanRRRR** [reply]
Tom,
I though to use your way of conversion but i know it is slow (1000x slower than unformatted Fortran), we used it as a last resort sometimes with HDF5, where there exist the analog of sdffilter called h5dump or our own Python HDF converter. By the way do you or anybody have these utilities compiled for Windows? At least they will be compatible with all  version of Windows. I'd try them, may be somehow miraculously they are fast. 

I do not get how billions files appeared here? You have one SDF files per intermediate output, in Fortran approach there will be better to separate particles and fields and there will be 2 files. Some other PIC codes have 5-10 files per intermediate output, and at least for myself i find that optimal (electrons and each type of atoms saved also separately).

As to parallel vs serial speed - some codes may need parallel I/O of course. But tasks  with up to 50 TB of total output per run this Fortran approach will fit perfectly. Specifically if at some point you plan to have your personal supercomputer at home :)

**Reply by TomGoffrey** [reply]
The billions of files would occur because the code you've written would be executed by every single process, each writing their own file. You have 100,000 MPI processes running, and each output 10,000 files (extreme, but not unheard of). That gives you a 1e9 files.

One final thing to consider - it's not the total size of the data that is necessarily going to limit if you need to run in parallel. More likely the amount of compute you have to do. Even for 2D simulations we typically require >100,000 processes to carry out our calculations in a reasonable time frame.

**Reply by DanRRRR** [reply]
I understand. If SDF or any format is used for intermediate scratch data let the code use it as usually. I am talking only about data for final saving to keep the data.

**Reply by DanRRRR** [reply]
By the way the new standard PCIe 5 also has 16 GB/s bandwidth. So soon every laptop will get the bandwidth of that superduper ARCHER.

**Comment by DanRRRR** [comment]
I just contacted the developer's forums of Visit and Paraview. They quickly responded that they could be interested to work on developing the ability of their software to read SDF file format. I am surprised why no one contacted them before. They ask where to find the description of SDF ? What one can read in EPOCH description is the following "A detailed specification of this format is available elsewhere, although this is only of interest to developers wishing to write new libraries". Is Self Describing Format the same as SDFormat which also exists in the world wild net confusing things even more?

**Reply by Status-Mirror** [reply]
Hi Dan,

I'm not an SDF developer, so I can't give you a complete answer - but you may find this helpful. The SDF documentation in .tex form can be found here:
https://github.com/Warwick-Plasma/SDF_documentation/blob/main/sdf_format.tex

Hope this helps,
Stuart

**Comment by DanRRRR** [comment]
The maintainer of Visit Kathleen Biagas asking : "Since EPOCH has a [VisIt plugin for SDF](https://epochpic.github.io/documentation/visualising_output/visualising_sdf_files_with_llnl_visit.html) already, I wonder if the EPOCH folks would be amenable to contributing their plugin to VisIt's repo? I don't know if there would be licensing issues or whatnot"

Probably EPOCH developers should be very interested to give the users the ability to visualize the output of the code straight off the run and not through the several more intermediate stages of different software you need to learn to install and use and which may or may not work anyway.
